cli/flox/src/commands/publish.rs

```diff
@@ -0,0 +1,126 @@
+use crate::config::Config;
+use crate::subcommand_metric;
+use crate::utils::message;
+
+#[allow(unused)] // remove when we implement the command
+#[derive(Bpaf, Clone)]
+pub struct Publish {
+    #[bpaf(external(environment_select), fallback(Default::default()))]
+    environment: EnvironmentSelect,
+
+    #[bpaf(external(subcommand_or_publish_target))]
+    subcommand_or_publish_target: SubcommandOrPublishTarget,
+}
+
+#[derive(Debug, Bpaf, Clone)]
+enum SubcommandOrPublishTarget {
+    PublishTarget {
+        /// The package to build.
+        /// Corresponds to entries in the 'build' table in the environment's manifest.toml.
+        /// If not specified, all packages are built.
+        #[bpaf(positional("package"))]
+        target: String,
+    },
+}
+
+impl Publish {
+    pub async fn handle(self, config: Config, flox: Flox) -> Result<()> {
+        if !config.features.unwrap_or_default().publish {
+            message::plain("🚧 👷 heja, a new command is in construction here, stay tuned!");
+            bail!("'publish' feature is not enabled.");
+        }
+
+        match self.subcommand_or_publish_target {
+            SubcommandOrPublishTarget::PublishTarget { target } => {
+                let env = self
+                    .environment
+                    .detect_concrete_environment(&flox, "Publish")?;
+
+                Self::publish(flox, env, target).await
+            },
+        }
+    }
+
+    #[instrument(name = "publish", skip_all, fields(package))]
+    async fn publish(mut flox: Flox, mut env: ConcreteEnvironment, package: String) -> Result<()> {
+        subcommand_metric!("publish");
+
+        if !check_package(&env.dyn_environment_ref_mut().lockfile(&flox)?, &package)? {
+            bail!("Package '{}' not found in environment", package);
+        }
+
+        let env_metadata = match &mut env {
+            ConcreteEnvironment::Managed(environment) => {
+                check_environment_metadata(&flox, environment)
+            },
+            ConcreteEnvironment::Path(environment) => {
+                check_environment_metadata(&flox, environment)
+            },
+            _ => bail!("Unsupported environment type"),
+        }
+        .or_else(|e| bail!(e.to_string()))?;
+
+        let build_metadata = match &env {
+            ConcreteEnvironment::Managed(environment) => {
+                check_build_metadata(environment, &package, &flox.system)
+            },
+            ConcreteEnvironment::Path(environment) => {
+                check_build_metadata(environment, &package, &flox.system)
+            },
+            _ => unreachable!(),
+        }
+        .or_else(|e| bail!(e.to_string()))?;
+
+        let publish_provider = PublishProvider::<&FloxBuildMk> {
+            build_metadata,
+            env_metadata,
+            _builder: None,
+        };
+
+        ensure_floxhub_token(&mut flox).await?;
+        let token = flox.floxhub_token.as_ref().unwrap();
```

Comment:
(non-blocking) Should never hit this, but just in case the behaviour of `ensure_floxhub_token` changes. Sidenote: maybe it should actually return the token?
```suggestion
        let token = flox
            .floxhub_token
            .as_ref()
            .expect("should be authenticated to FloxHub");
```

Reply:
Good idea.  I have a few follow up items for later tickets (like also using the `create_test_token` elsewhere).  Added it to that list and took your suggestion.